crates/asm/common/src/aux/provider.rs

```diff
@@ -0,0 +1,188 @@
+//! Auxiliary data provider.
+//!
+//! Provides verified auxiliary data to subprotocols during the processing phase.
+
+use std::collections::HashMap;
+
+use bitcoin::{Transaction, Txid};
+use strata_merkle::CompactMmr64;
+
+use crate::{AsmHasher, AuxError, AuxResult, Hash32, aux::data::AuxData};
+
+/// Provides verified auxiliary data to subprotocols during transaction processing.
+///
+/// The provider verifies all auxiliary data upfront during construction and stores
+/// it in efficient lookup structures for O(1) access:
+///
+/// - **Bitcoin transactions**: Decoded and indexed by txid in a hashmap
+/// - **Manifest leaves**: MMR proofs verified and indexed by MMR position
+///
+/// All verification happens during construction via [`try_new`](Self::try_new), so
+/// subsequent getter method calls return already-verified data without additional
+/// validation overhead.
+#[derive(Debug, Clone)]
+pub struct AuxDataProvider {
+    /// Verified Bitcoin transactions indexed by txid
+    txs: HashMap<Txid, Transaction>,
+    /// Verified manifest leaves indexed by MMR index
+    manifest_leaves: HashMap<u64, Hash32>,
+}
+
+impl AuxDataProvider {
+    /// Attempts to create a new provider by verifying and indexing all auxiliary data.
+    ///
+    /// Decodes and verifies all Bitcoin transactions and manifest leaves from the provided
+    /// unverified data. If any verification fails, returns an error and no provider is created.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - Unverified auxiliary data containing Bitcoin transactions and manifest leaves
+    /// * `compact_mmr` - Compact MMR snapshot used to verify manifest leaf proofs
+    ///
+    /// # Errors
+    ///
+    /// Returns `AuxError::InvalidBitcoinTx` if any transaction fails to decode or is malformed.
+    /// Returns `AuxError::InvalidMmrProof` if any manifest leaf's MMR proof fails verification.
+    pub fn try_new(data: &AuxData, compact_mmr: &CompactMmr64<[u8; 32]>) -> AuxResult<Self> {
```

Comment:
This acts like a constructor but actually does a lot more complicated validation and stuff.  This should be broken out into a separate function that does these checks and then calls an actual constructor.